src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1494,16 +1494,44 @@ internal void AssertMemberExposure(Symbol member, bool forDiagnostics = false)
                 member = e;
             }
 
-            var declared = Volatile.Read(ref _lazyDeclaredMembersAndInitializers);
-            RoslynDebug.AssertOrFailFast(declared != DeclaredMembersAndInitializers.UninitializedSentinel);
+            var membersAndInitializers = Volatile.Read(ref _lazyMembersAndInitializers);
 
-            if ((declared is object && (declared.NonTypeMembers.Contains(m => m == (object)member) || declared.RecordPrimaryConstructor == (object)member)) ||
-                Volatile.Read(ref _lazyMembersAndInitializers)?.NonTypeMembers.Contains(m => m == (object)member) == true)
+            if (isMemberInCompleteMemberList(membersAndInitializers, member))
             {
                 return;
             }
 
+            if (membersAndInitializers is null)
+            {
+                var declared = Volatile.Read(ref _lazyDeclaredMembersAndInitializers);
+                RoslynDebug.AssertOrFailFast(declared != DeclaredMembersAndInitializers.UninitializedSentinel);
+
+                if (declared is object)
+                {
+                    if (declared.NonTypeMembers.Contains(m => m == (object)member) || declared.RecordPrimaryConstructor == (object)member)
+                    {
+                        return;
+                    }
+                }
+                else
+                {
+                    // It looks like there was a race and we need to check _lazyMembersAndInitializers again
+                    membersAndInitializers = Volatile.Read(ref _lazyMembersAndInitializers);
```

Comment:
In order for this to be safe I believe we need to switch the following assignment to use a `Volatile.Write` call 

http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Source/SourceMemberContainerSymbol.cs,1458

```c#
var alreadyKnown = Interlocked.CompareExchange(ref _lazyMembersAndInitializers, membersAndInitializers, null);
if (alreadyKnown != null)
{
    diagnostics.Free();
    return alreadyKnown;
}

AddDeclarationDiagnostics(diagnostics);
diagnostics.Free();

// SUGGESTION: make this a Volatile.Write
_lazyDeclaredMembersAndInitializers = null;

return membersAndInitializers!;
```

The `CompareExchange(ref _lazyMembersAndInitializes, ...)` uses a fully memory barrier but I _believe_ that occurs before the operation not after. That means all of the writes after the operation can be viewed by other thread in any order. Hence it's possible for a different thread to observe `_lazyDeclaredMembersAndInitializers = null` before it sees the write to `_lazyMembersAndInitiaizers`. That is true even though you use `Volatile.Read` because that only affects reordering on the current thread. 

Rule of thumb is that if you use `Volatile.Read` on one thread they need to be paired with `Volatile.Write` or `Intelocked.Exchange` on the other thread. Else you can get into ordering issues. 

I'm not 100% confident in my assessment here, memory ordering is tricky. So possible someone comes back and corrects me here but would feel better with this change.

Reply:
Followed up with runtime team. My reasoning is incorrect. The `CompareExchange` does a fence before / after hence the writes are ordered.